File: android/src/main/kotlin/com/example/shipbook/ShipbookPlugin.kt
package com.example.shipbook

import android.app.Application
import android.content.Context
import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.shipbook.shipbooksdk.Log
import io.shipbook.shipbooksdk.ShipBook

/** ShipbookPlugin */
class ShipbookPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  private lateinit var context: Context

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    this.context = flutterPluginBinding.applicationContext
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "shipbook")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
      "start" -> start(call, result)
      "e" -> e(call, result)
      "w" -> w(call, result)
      "i" -> i(call, result)
      "d" -> d(call, result)
      "v" -> v(call, result)
      else -> {
        result.notImplemented()
      }
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  private fun start(@NonNull call: MethodCall, @NonNull result: Result) {
    var arguments = call.arguments as Map<String, Any>
    var appId = arguments["appId"] as String
    var appKey = arguments["appKey"] as String
    ShipBook.start(context as Application, appId, appKey);
    result.success(null)
  }

  private fun e(@NonNull call: MethodCall, @NonNull result: Result) {
    var arguments = call.arguments as Map<String, Any>
    var tag = arguments["tag"] as String
    var message = arguments["message"] as String
    Log.e(tag, message);
    result.success(null)
  }

  private fun w(@NonNull call: MethodCall, @NonNull result: Result) {
    var arguments = call.arguments as Map<String, Any>
    var tag = arguments["tag"] as String
    var message = arguments["message"] as String
    Log.w(tag, message);
    result.success(null)
  }

  private fun i(@NonNull call: MethodCall, @NonNull result: Result) {
    var arguments = call.arguments as Map<String, Any>
    var tag = arguments["tag"] as String
    var message = arguments["message"] as String
    Log.i(tag, message);
    result.success(null)
  }

  private fun d(@NonNull call: MethodCall, @NonNull result: Result) {
    var arguments = call.arguments as Map<String, Any>
    var tag = arguments["tag"] as String
    var message = arguments["message"] as String
    Log.d(tag, message);
    result.success(null)
  }

  private fun v(@NonNull call: MethodCall, @NonNull result: Result) {
    var arguments = call.arguments as Map<String, Any>
    var tag = arguments["tag"] as String
    var message = arguments["message"] as String
    Log.v(tag, message);
    result.success(null)
  }
}

File: ios/Classes/ShipbookPlugin.swift
import Flutter
import UIKit
import ShipBookSDK

public class ShipbookPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "shipbook", binaryMessenger: registrar.messenger())
    let instance = ShipbookPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
        result("iOS " + UIDevice.current.systemVersion)
    case "start":
        start(call, result)
    case "e":
        e(call, result)
    case "w":
        w(call, result)
    case "i":
        i(call, result)
    case "d":
        d(call, result)
    case "v":
        v(call, result)
    default:
        result(nil)
    }
  }

  private func start(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
      let arguments = call.arguments as! Dictionary<String, Any>
      let appId = arguments["appId"] as! String;
      let appKey = arguments["appKey"] as! String;
      ShipBook.start(appId: appId, appKey: appKey);
      result(nil)
  }

  private func e(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
      let arguments = call.arguments as! Dictionary<String, Any>
      let tag = arguments["tag"] as! String;
      let message = arguments["message"] as! String;

      let log = ShipBook.getLogger(tag);
      log.e(message);
      result(nil)
  }

  private func w(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
      let arguments = call.arguments as! Dictionary<String, Any>
      let tag = arguments["tag"] as! String;
      let message = arguments["message"] as! String;

      let log = ShipBook.getLogger(tag);
      log.w(message);
      result(nil)
  }

  private func i(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
      let arguments = call.arguments as! Dictionary<String, Any>
      let tag = arguments["tag"] as! String;
      let message = arguments["message"] as! String;

      let log = ShipBook.getLogger(tag);
      log.i(message);
      result(nil)
  }

  private func d(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
      let arguments = call.arguments as! Dictionary<String, Any>
      let tag = arguments["tag"] as! String;
      let message = arguments["message"] as! String;

      let log = ShipBook.getLogger(tag);
      log.d(message);
      result(nil)
  }

  private func v(_ call: FlutterMethodCall,_ result: @escaping FlutterResult) {
      let arguments = call.arguments as! Dictionary<String, Any>
      let tag = arguments["tag"] as! String;
      let message = arguments["message"] as! String;

      let log = ShipBook.getLogger(tag);
      log.v(message);
      result(nil)
  }
}
